pkg/client/cache/watch_cache.go

```diff
@@ -0,0 +1,255 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"sync"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// watchCacheElement is a single "watch event" stored in a cache.
+// It contains the resource version of the object and the object
+// itself.
+type watchCacheElement struct {
+	resourceVersion uint64
+	event           watch.Event
+}
+
+// WatchCache implements a Store interface.
+// However, it depends on the elements implementing runtime.Object interface.
+//
+// WatchCache is a "sliding window" (with a limitted capacity) of objects
+// observed from a watch.
+type WatchCache struct {
+	sync.RWMutex
+
+	// Maximum size of history window.
+	capacity int
+
+	// cache is used a cyclic buffer - its first element (with the smallest
+	// resourceVersion) is defined by startIndex, its last element is defined
+	// by endIndex (if cache is full it will be startIndex + capacity).
+	// Both startIndex and endIndex can be greater than buffer capacity -
+	// you should always apply modulo capacity to get an index in cache array.
+	cache      []watchCacheElement
+	startIndex int
+	endIndex   int
+
+	// store will effectively support LIST operation from the "end of cache
+	// history" i.e. from the moment just after the newest cached watched event.
+	// It is necessary to effectively allow clients to start watching at now.
+	store Store
+
+	// ResourceVersion up to which the WatchCache is propagated.
+	resourceVersion uint64
+
+	// This handler is run at the end of every successful Replace() method.
+	onReplace func()
+
+	// This handler is run at the end of every Add/Update/Delete method.
+	onEvent func(watch.Event)
+}
+
+func NewWatchCache(capacity int) *WatchCache {
+	return &WatchCache{
+		capacity:        capacity,
+		cache:           make([]watchCacheElement, capacity),
+		startIndex:      0,
+		endIndex:        0,
+		store:           NewStore(MetaNamespaceKeyFunc),
+		resourceVersion: 0,
+	}
+}
+
+func (w *WatchCache) Add(obj interface{}) error {
+	object, ok := obj.(runtime.Object)
+	if !ok {
+		return fmt.Errorf("obj does not implement runtime.Object interface: %v", obj)
+	}
+	meta, err := meta.Accessor(object)
+	if err != nil {
+		return err
+	}
+	resourceVersion, err := strconv.ParseUint(meta.ResourceVersion(), 10, 64)
+	if err != nil {
+		return err
+	}
+	event := watch.Event{watch.Added, object}
+
+	w.Lock()
+	defer w.Unlock()
+	if w.onEvent != nil {
+		w.onEvent(event)
+	}
+	w.updateCache(resourceVersion, event)
+	w.resourceVersion = resourceVersion
+	return w.store.Add(object)
+}
+
+func (w *WatchCache) Update(obj interface{}) error {
```

Comment:
It looks like a lot of the code in Add, Update, and Delete is duplicated. I'd suggest factoring that code out into a separate function. I think it's fine to leave that for a subsequent PR, though.